src/webgl-preview.ts

```diff
@@ -423,130 +331,21 @@ export class WebGLPreview {
   private renderFrame(layerCount: number): void {
     this.group = this.createGroup('layer' + this.renderLayerIndex);
 
-    for (let l = 0; l < layerCount && this.renderLayerIndex + l < this.layers.length; l++) {
-      this.renderLayer(this.renderLayerIndex);
-      this.renderLayerIndex++;
-    }
-
-    this.batchGeometries();
-
-    this.scene.add(this.group);
-  }
-
-  /**
-   *  @internal
-   */
-  renderLayer(index: number): void {
-    if (index > this.maxLayerIndex) return;
-    const l = this.layers[index];
-
-    const currentLayer: RenderLayer = {
-      extrusion: [],
-      travel: [],
-      z: this.state.z,
-      height: l.height
-    };
-
-    for (const cmd of l.commands) {
-      if (cmd.gcode == 'g20') {
-        this.setInches();
-        continue;
-      }
-
-      if (cmd.gcode.startsWith('t')) {
-        // flush render queue
-        this.doRenderExtrusion(currentLayer, index);
-        currentLayer.extrusion = [];
-
-        const tool = cmd as SelectToolCommand;
-        this.state.t = tool.toolIndex;
-        continue;
-      }
-
-      if (['g0', 'g00', 'g1', 'g01', 'g2', 'g02', 'g3', 'g03'].indexOf(cmd.gcode) > -1) {
-        const g = cmd as MoveCommand;
-        const next: State = {
-          x: g.params.x ?? this.state.x,
-          y: g.params.y ?? this.state.y,
-          z: g.params.z ?? this.state.z,
-          r: g.params.r ?? this.state.r,
-          e: g.params.e ?? this.state.e,
-          i: g.params.i ?? this.state.i,
-          j: g.params.j ?? this.state.j,
-          t: this.state.t
-        };
-
-        if (index >= this.minLayerIndex) {
-          const extrude = (g.params.e ?? 0) > 0 || this.nonTravelmoves.indexOf(cmd.gcode) > -1;
-          const moving = next.x != this.state.x || next.y != this.state.y || next.z != this.state.z;
-          if (moving) {
-            if ((extrude && this.renderExtrusion) || (!extrude && this.renderTravel)) {
-              if (cmd.gcode == 'g2' || cmd.gcode == 'g3' || cmd.gcode == 'g02' || cmd.gcode == 'g03') {
-                this.addArcSegment(currentLayer, this.state, next, extrude, cmd.gcode == 'g2' || cmd.gcode == 'g02');
-              } else {
-                this.addLineSegment(currentLayer, this.state, next, extrude);
-              }
-            }
-          }
-        }
-
-        // update this.state
-        this.state.x = next.x;
-        this.state.y = next.y;
-        this.state.z = next.z;
-        // if (next.e) state.e = next.e; // where not really tracking e as distance (yet) but we only check if some commands are extruding (positive e)
-        if (!this.beyondFirstMove) this.beyondFirstMove = true;
-      }
-    }
-
-    this.doRenderExtrusion(currentLayer, index);
-  }
+    const endIndex = Math.min(this.renderLayerIndex + layerCount, this.job.layers().length - 1);
+    const layersToRender = this.job
```

Comment:
isn't this a list of paths, now that you've flattened the layer? suggestion to name it pathsToRender